File: blog-web/src/types/post.ts
import type { Category } from './category'
import type { Tag } from './tag'
import type { Comment } from './comment'

export interface Post {
  id: string
  title: string
  slug: string
  excerpt: string
  coverImage?: string
  viewCount: number
  likeCount: number
  published: boolean
  pinned: boolean
  createdAt: string
  updatedAt: string
  category: Category
  tags: Tag[]
}

export interface PostDetail extends Post {
  content: string
  contentHtml: string
  comments: Comment[]
}

export interface PostCreateRequest {
  title: string
  slug?: string
  content: string
  contentHtml: string
  excerpt?: string
  coverImage?: string
  categoryId?: string
  tagIds?: string[]
  published: boolean
  pinned: boolean
}

export interface PostUpdateRequest extends PostCreateRequest {
  id: string
}

File: blog-server/src/main/java/com/pidan/blog/entity/Post.java
package com.pidan.blog.entity;

import com.pidan.blog.common.BaseEntity;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.*;

@Entity
@Table(name = "posts")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Post extends BaseEntity {

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, unique = true, length = 200)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String content;

    @Column(name = "content_html", columnDefinition = "TEXT")
    private String contentHtml;

    @Column(name = "cover_image")
    private String coverImage;

    @Column(name = "view_count")
    @Builder.Default
    private Long viewCount = 0L;

    @Column(name = "like_count")
    @Builder.Default
    private Long likeCount = 0L;

    @Column(nullable = false)
    @Builder.Default
    private Boolean published = false;

    @Column(nullable = false)
    @Builder.Default
    private Boolean pinned = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "post_tags",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @Builder.Default
    private Set<Tag> tags = new HashSet<>();

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Comment> comments = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
